File: src-tauri/src/analysis/tool_analytics/scanner.rs
use std::collections::HashMap;
use std::io::{BufRead, BufReader};
use std::path::Path;

use chrono::{Datelike, Local, TimeZone, Timelike};

use crate::analysis::tokenizer;

use super::shared::{
    heatmap_key, parse_timestamp_ms, tool_result_text, HeatmapCellAcc, HeatmapKey, RawEntry,
    ToolCallStart, ToolStats,
};

/// Scan a single JSONL session, folding tool_use/tool_result pairs into `stats`.
pub(super) fn scan_session(path: &Path, stats: &mut HashMap<String, ToolStats>) -> Option<()> {
    let file = std::fs::File::open(path).ok()?;
    let reader = BufReader::with_capacity(64 * 1024, file);
    let mut in_flight: HashMap<String, ToolCallStart> = HashMap::new();

    for line in reader.lines().map_while(Result::ok) {
        if line.trim().is_empty() {
            continue;
        }
        let entry: RawEntry = match serde_json::from_str(&line) {
            Ok(e) => e,
            Err(_) => continue,
        };
        let ts_ms = entry
            .timestamp
            .as_deref()
            .and_then(parse_timestamp_ms)
            .unwrap_or(0.0);
        let msg = match entry.message {
            Some(m) => m,
            None => continue,
        };
        let content = match msg.content {
            Some(c) => c,
            None => continue,
        };
        let blocks = match content.as_array() {
            Some(a) => a,
            None => continue,
        };

        match msg.role.as_deref() {
            Some("assistant") => {
                for block in blocks {
                    if block.get("type").and_then(|v| v.as_str()) != Some("tool_use") {
                        continue;
                    }
                    let id = match block.get("id").and_then(|v| v.as_str()) {
                        Some(s) => s.to_string(),
                        None => continue,
                    };
                    let name = block
                        .get("name")
                        .and_then(|v| v.as_str())
                        .unwrap_or("unknown")
                        .to_string();
                    in_flight.insert(
                        id,
                        ToolCallStart {
                            tool_name: name,
                            start_ms: ts_ms,
                        },
                    );
                }
            }
            Some("user") => {
                for block in blocks {
                    if block.get("type").and_then(|v| v.as_str()) != Some("tool_result") {
                        continue;
                    }
                    let id = match block.get("tool_use_id").and_then(|v| v.as_str()) {
                        Some(s) => s.to_string(),
                        None => continue,
                    };
                    let call = match in_flight.remove(&id) {
                        Some(c) => c,
                        None => continue,
                    };
                    let is_error = block.get("is_error").and_then(|v| v.as_bool()).unwrap_or(false);
                    let result_text = block
                        .get("content")
                        .map(tool_result_text)
                        .unwrap_or_default();
                    let token_count = tokenizer::count_tokens(&result_text) as u64;
                    let duration = (ts_ms - call.start_ms).max(0.0);

                    let entry = stats.entry(call.tool_name).or_default();
                    entry.call_count += 1;
                    if is_error {
                        entry.error_count += 1;
                    } else {
                        entry.success_count += 1;
                    }
                    if duration > 0.0 {
                        entry.duration_samples.push(duration);
                    }
                    entry.token_samples.push(token_count);
                }
            }
            _ => {}
        }
    }
    Some(())
}

pub(super) fn bucket_local(ts_ms: f64) -> Option<(u8, u8)> {
    let secs = (ts_ms / 1000.0) as i64;
    let nanos = ((ts_ms - (secs as f64) * 1000.0) * 1_000_000.0).round() as u32;
    let utc = chrono::DateTime::from_timestamp(secs, nanos)?;
    let local = Local.from_utc_datetime(&utc.naive_utc());
    let day = local.weekday().num_days_from_monday() as u8;
    let hour = local.hour() as u8;
    Some((day, hour))
}

/// Walk a session's tool_use blocks, bucketing by local (weekday, hour). When
/// `tool_filter` is `Some`, only matching tool names count.
pub(super) fn scan_session_heatmap(
    path: &Path,
    buckets: &mut HashMap<HeatmapKey, HeatmapCellAcc>,
    tool_filter: Option<&str>,
) -> Option<()> {
    let file = std::fs::File::open(path).ok()?;
    let reader = BufReader::with_capacity(64 * 1024, file);

    for line in reader.lines().map_while(Result::ok) {
        if line.trim().is_empty() {
            continue;
        }
        let entry: RawEntry = match serde_json::from_str(&line) {
            Ok(e) => e,
            Err(_) => continue,
        };
        let ts_ms = match entry.timestamp.as_deref().and_then(parse_timestamp_ms) {
            Some(v) => v,
            None => continue,
        };
        let msg = match entry.message {
            Some(m) => m,
            None => continue,
        };
        if msg.role.as_deref() != Some("assistant") {
            continue;
        }
        let blocks = match msg.content.as_ref().and_then(|c| c.as_array()) {
            Some(a) => a,
            None => continue,
        };

        let (day, hour) = match bucket_local(ts_ms) {
            Some(pair) => pair,
            None => continue,
        };

        for block in blocks {
            if block.get("type").and_then(|v| v.as_str()) != Some("tool_use") {
                continue;
            }
            let name = block
                .get("name")
                .and_then(|v| v.as_str())
                .unwrap_or("unknown");
            if let Some(filter) = tool_filter {
                if name != filter {
                    continue;
                }
            }
            let cell = buckets.entry(heatmap_key(day, hour)).or_default();
            cell.total += 1;
            *cell.per_tool.entry(name.to_string()).or_insert(0) += 1;
        }
    }
    Some(())
}

File: internal/insights/tool_analytics/scanner.go
package tool_analytics

import (
	"bufio"
	"encoding/json"
	"os"
	"strings"
	"time"

	"claude-devtools/internal/tokenizer"
)

// scanSession scans a single JSONL session file and folds tool_use/tool_result
// pairs into stats. Mirrors scanner.rs::scan_session.
func scanSession(path string, stats map[string]*toolStats) {
	f, err := os.Open(path)
	if err != nil {
		return
	}
	defer f.Close()

	inFlight := make(map[string]toolCallStart)

	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 64*1024), 4*1024*1024)
	for scanner.Scan() {
		line := scanner.Text()
		if len(strings.TrimSpace(line)) == 0 {
			continue
		}

		var entry rawEntry
		if err := json.Unmarshal([]byte(line), &entry); err != nil {
			continue
		}

		tsMs := 0.0
		if entry.Timestamp != nil {
			if ms, ok := parseTimestampMs(*entry.Timestamp); ok {
				tsMs = ms
			}
		}

		if entry.Message == nil || entry.Message.Content == nil {
			continue
		}

		role := ""
		if entry.Message.Role != nil {
			role = *entry.Message.Role
		}

		var blocks []json.RawMessage
		if err := json.Unmarshal(*entry.Message.Content, &blocks); err != nil {
			continue
		}

		switch role {
		case "assistant":
			for _, b := range blocks {
				var block struct {
					Type string `json:"type"`
					ID   string `json:"id"`
					Name string `json:"name"`
				}
				if json.Unmarshal(b, &block) != nil || block.Type != "tool_use" {
					continue
				}
				id := block.ID
				if id == "" {
					continue
				}
				name := block.Name
				if name == "" {
					name = "unknown"
				}
				inFlight[id] = toolCallStart{toolName: name, startMs: tsMs}
			}

		case "user":
			for _, b := range blocks {
				var block struct {
					Type      string           `json:"type"`
					ToolUseID string           `json:"tool_use_id"`
					IsError   *bool            `json:"is_error"`
					Content   *json.RawMessage `json:"content"`
				}
				if json.Unmarshal(b, &block) != nil || block.Type != "tool_result" {
					continue
				}
				call, ok := inFlight[block.ToolUseID]
				if !ok {
					continue
				}
				delete(inFlight, block.ToolUseID)

				isError := block.IsError != nil && *block.IsError
				resultText := ""
				if block.Content != nil {
					resultText = toolResultText(*block.Content)
				}
				tokenCount := uint64(tokenizer.CountTokens(resultText))
				duration := tsMs - call.startMs
				if duration < 0 {
					duration = 0
				}

				st := stats[call.toolName]
				if st == nil {
					st = &toolStats{}
					stats[call.toolName] = st
				}
				st.callCount++
				if isError {
					st.errorCount++
				} else {
					st.successCount++
				}
				if duration > 0 {
					st.durationSamples = append(st.durationSamples, duration)
				}
				st.tokenSamples = append(st.tokenSamples, tokenCount)
			}
		}
	}
}

// bucketLocal converts a millisecond UTC timestamp to (weekday, hour) in local
// time. Mirrors scanner.rs::bucket_local.
// weekday: 0=Monday … 6=Sunday (matches chrono num_days_from_monday).
func bucketLocal(tsMs float64) (day, hour uint8, ok bool) {
	secs := int64(tsMs / 1000.0)
	t := time.Unix(secs, 0).Local()
	// Go Weekday: Sunday=0, Monday=1, …, Saturday=6.
	// Chrono num_days_from_monday: Monday=0, …, Sunday=6.
	goWd := int(t.Weekday()) // 0=Sun…6=Sat
	chromoDay := (goWd + 6) % 7 // convert: Mon=0,…,Sun=6
	return uint8(chromoDay), uint8(t.Hour()), true
}

// scanSessionHeatmap walks a session's assistant tool_use blocks, bucketing by
// local (weekday, hour). If toolFilter is non-empty, only matching tool names
// count. Mirrors scanner.rs::scan_session_heatmap.
func scanSessionHeatmap(path string, buckets map[heatmapKey]*heatmapCellAcc, toolFilter string) {
	f, err := os.Open(path)
	if err != nil {
		return
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 64*1024), 4*1024*1024)
	for scanner.Scan() {
		line := scanner.Text()
		if len(strings.TrimSpace(line)) == 0 {
			continue
		}

		var entry rawEntry
		if err := json.Unmarshal([]byte(line), &entry); err != nil {
			continue
		}
		if entry.Timestamp == nil {
			continue
		}
		tsMs, ok := parseTimestampMs(*entry.Timestamp)
		if !ok {
			continue
		}

		if entry.Message == nil || entry.Message.Content == nil {
			continue
		}
		if entry.Message.Role == nil || *entry.Message.Role != "assistant" {
			continue
		}

		var blocks []json.RawMessage
		if json.Unmarshal(*entry.Message.Content, &blocks) != nil {
			continue
		}

		day, hour, _ := bucketLocal(tsMs)

		for _, b := range blocks {
			var block struct {
				Type string `json:"type"`
				Name string `json:"name"`
			}
			if json.Unmarshal(b, &block) != nil || block.Type != "tool_use" {
				continue
			}
			name := block.Name
			if name == "" {
				name = "unknown"
			}
			if toolFilter != "" && name != toolFilter {
				continue
			}
			key := makeHeatmapKey(day, hour)
			cell := buckets[key]
			if cell == nil {
				cell = &heatmapCellAcc{perTool: make(map[string]uint32)}
				buckets[key] = cell
			}
			cell.total++
			cell.perTool[name]++
		}
	}
}
